threading.h

```diff
@@ -92,13 +95,15 @@ namespace tgvoip{
 		static void* ActualEntryPoint(void* arg){
 			Thread* self=reinterpret_cast<Thread*>(arg);
 			if(self->name){
-#if !defined(__APPLE__) && !defined(__gnu_hurd__)
+#if defined(__linux__) || defined(__FreeBSD__)
 				pthread_setname_np(self->thread, self->name);
-#elif !defined(__gnu_hurd__)
+#elif defined(__APPLE__)
```

Comment:
And here too

Reply:
> @HenryHu seem to replace the general ifndef's with more specific positive ones. I don't think this is the right way to go. I think this should be removed from the patchset and the new ifdef's should be as more generic as they can be, currently this feels like a couple of hacks here and there.

Are we talking about first two changes in the queue e21a219 and ccd3a0c? The use of `__FreeBSD_kernel__` is not correct here. First, it is not visible, cause it comes from `<param.h>` that libtgvoip doesn't include, while `__FreeBSD__` is provided by the compiler. Second, this define is provided for software that uses certain kernel features and it is not recommended for generic portable code. Quoting param.h:
```
/*
 * __FreeBSD_kernel__ indicates that this system uses the kernel of FreeBSD,
 * which by definition is always true on FreeBSD. This macro is also defined 
 * on other systems that use the kernel of FreeBSD, such as GNU/kFreeBSD. 
 * 
 * It is tempting to use this macro in userland code when we want to enable  
 * kernel-specific routines, and in fact it's fine to do this in code that  
 * is part of FreeBSD itself.  However, be aware that as presence of this
 * macro is still not widespread (e.g. older FreeBSD versions, 3rd party
 * compilers, etc), it is STRONGLY DISCOURAGED to check for this macro in
 * external applications without also checking for __FreeBSD__ as an
 * alternative.
 */

```

Reply:
No, I'm talking about the fact you're using `__FreeBSD__` instead of some unix- or posix-related define

Reply:
> No, I'm talking about the fact you're using `__FreeBSD__` instead of some unix- or posix-related define

Because the code being enabled is neither UNIX nor POSIX. Specifically, the defines we are talking about enable ALSA/Pulse compilation. These projects have explicit list of supported OSes, and of course not all OSes that define `__unix__` are in this list.

Also, this goes together with the existing style. If we look at a wider scope, the whole libtgvoip uses explicit specific defines: `__linux__`, `__gnu__hurd__`, `__APPLE__`, `__ANDROID__` and `_WIN32`. The only exception is in threading.h, and for a good reason - cause it looks for a POSIX feature. I absolutely agree with this style, because if we just blindly throw `__unix__` instead of the explicit list, the chances that we enable successful compilation on any additional operating system are so minuscule that I would bet $1000 against that.

However, if you insist, I can substitute `__linux__ || __FreeBSD__` to `__unix__`. That would enable compilation for me, still compiling for you on Linux. This would need extra review and testing to avoid breaking OS X compilation, cause it defines `__unix__`, but libtgvoip treats it differently to Linux.

Reply:
I don't want you to blindly replace all `__linux__ || __FreeBSD__` to `__unix`, I want to hear which functions belong to which standard and which ifdef's actually should be used. So I googled pthread_setname_np on my own and found this:
![изображение](https://user-images.githubusercontent.com/17829319/133490653-8fc95fbc-bfe2-4045-a424-b5266f8ee2aa.png)
Now I want to hear which OSes actually have this extension and what is the right way to check for them.

Reply:
I lost the context. You started with

> HenryHu seem to replace the general ifndef's with more specific positive ones

And I was replying to that, hence mentioning ALSA/Pulse and why it is not UNIX feature.

Then you switched back to non-portable pthread extensions, which I believed, we already discussed.

> I want to hear which functions belong to which standard and which ifdef's actually should be used.
>  Now I want to hear which OSes actually have this extension and what is the right way to check for them.

Very good. So, this is what exactly what my patch does. It provides explicit disjunct that guarantees presence of pthread_setname_np(). With such positive logic the code is guaranteed to compile. The previous logic `#if !defined(__APPLE__) && !defined(__gnu_hurd__)` was flawed. Not being Apple or Hurd doesn't guarantee you presence of pthread_setname_np().

Reply:
Haiku doesn't have pthread_setname_np(), they have rename_thread() [Example code](https://github.com/haikuports/haikuports/pull/4534/commits/aa46a627cfd926ad203c0746740a38e6de1bc6e9#diff-bdea86ab3034fb2e035c4f7a860b2d74152859ff5e5da63210de01ba85a32d15R62)

My change would reduce number of compilation errors on Haiku by one, cause new ifdef logic will not try to compile pthread_setname_np() on Haiku, while old does.

I'd suggest not to mix Haiku and FreeBSD in one pull request.